homeassistant/components/rainmachine/switch.py

```diff
@@ -287,6 +343,15 @@ async def async_enable_zone(self, *, zone_id: int) -> None:
         await self._controller.zones.enable(zone_id)
         await async_update_programs_and_zones(self.hass, self._entry)
 
+    async def async_push_weather_data(self, **weather_data: Any) -> None:
+        """Push weather data to the device."""
+        payload = {
+            ATTR_WEATHER: [
+                {key: value for key, value in weather_data.items() if value is not None}
+            ]
+        }
+        await self._controller.parsers.post_data(payload)
```

Comment:
Does each switch represent one controller? Otherwise the service should probably not be registered on the entity level.

Reply:
@MartinHjelmare A controller can have multiple switches (programs and zones) and one switch belongs to one controller. This seems like an entity service

Reply:
No, if more then one entity maps to the same controller it's not clear to the user to require an entity id to call the service. How do we explain that they need to use a switch entity and not another domain entity like sensor? Entity services should only be used when the feature is specific to a single entity.

If there's one device entry per controller we can use a device selector (device id) to identify the controller from the service call.

Example:
https://github.com/home-assistant/core/blob/1a7a4c52f12e764ca67895266f076468172e9f0f/homeassistant/components/system_bridge/services.yaml#L1-L11

https://github.com/home-assistant/core/blob/1a7a4c52f12e764ca67895266f076468172e9f0f/homeassistant/components/system_bridge/__init__.py#L115-L133

Reply:
@MartinHjelmare Okay – that must be new since this integration was developed (at which time, I was instructed to do entity services on "an entity in the domain" because that was the only way to associate a service with a particular device). This is why `pause_watering` and `unpause_watering` are registered under switch.

I'm hesitant to suggest this bigger change for this PR, which is already following a pre-established pattern. I would suggest a fast follow-up to this PR that makes this change more holistically.

Reply:
I'd recommend doing the rework first, and then add this PR correct from the beginning.

Reply:
Sounds good. @shbatm Let me get that done here shortly.

Reply:
https://github.com/home-assistant/core/pull/57145 – @shbatm, let me work with @MartinHjelmare to review/merge that one, then we can rebase this one.

Reply:
Ok. I agree w the plan to update first and then I'll rebase.